### DISCUSSION TITLE
Commit messages created using web interface use CRLF instead of LF

### DISCUSSION THREAD

**Discussion by tom93** [community/community#84731] [Pull Requests]
### Select Topic Area

Bug

### Body

Hi, please change the line endings of commit messages entered using the web interface from CRLF to LF.

Details:

Currently if I edit a commit message through the web interface, for example when I use "Squash and merge" on a pull request, the message of the resulting commit contains some CRLF line endings (carriage return + newline, aka Windows style). But commit messages in Git by convention use LF line endings (aka Unix style). I can't find an authoritative reference for this, but I'm sure that if you check virtually any repository (e.g. [git](https://github.com/git/git), [linux](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git), [github/docs](https://github.com/github/docs)) you'll find the commit messages use LF line endings.

Another related issue is that the messages of commits created using the web interface don't have a trailing newline (git automatically appends a trailing newline e.g. when using `git -m "..."`).

---

Full steps to reproduce:

1. Create a new repository (https://github.com/new), tick "Add a README file". My demo is: https://github.com/tom93/test-github-lineending

2. Clone the repository (I used `git clone git@github.com:tom93/test-github-lineending.git; cd test-github-lineending`)

3. Create a branch with a commit (the following commands assume bash):

```sh
git checkout -b branch
echo "test" >> README.md
git commit -a -m "First line

Second line
Third line"
```
4. Check that the commit message does not contain carriage returns:
`git show -s --format=format:%B branch | cat -v`
Expected output:

```
First line

Second line
Third line
```

6. Push using `git push origin branch` and create a pull request to merge the branch into main

7. Select the text "Second line<newline>Third line" from the first comment in the pull request stream, and copy that text.

8. In the "Merge pull request" dropdown, select "Squash and merge".

9. In the text area for setting the body of the sqaush commit, delete the text "Second line <newline>Third line" and paste the text from the clipboard (the pasted text will appear the same as the deleted text, but should have different line endings).

10. Press "Confirm squash and merge".

11. Fetch and inspect the line endings of the squashed commit:
`git fetch; git show -s --format=format:%B origin/main | cat -v`
Output:

```
First line (#1)

Second line^M
Third line
```

The ^M indicates that there is a carriage return after "Second line". In other words, the line ending is CRLF.

Variations:
A. If in step 6 I type the message by hand instead of pasting, I also get a carriage return in the message of the squashed commit.
B. If I skip step 6 (i.e. don't edit the text area), the message of the squashed commit does *not* contain a carriage return, as desired.

**Comment by georgetoloraia** [comment]
This is a test answer

**Comment by noahtallen** [comment]
This is actually more problematic than it sounds at first! `git subtree` commands grep the commit history to tell if a directory is a subtree:

https://github.com/noahtallen/git/blob/235986be822c9f8689be2e9a0b7804d0b1b6d821/contrib/subtree/git-subtree.sh#L420

When the line endings are mixed, the grep command fails. So if you do something like this:
1. Add a local subtree
2. Push the new branch to GitHub
3. Squash merge

Git will never see this directory as a subtree locally for anyone else who clones the repo. Because the command searching for `git-subtree-dir` in the commit history tries to match the line ending like this: `$dir/*\$`, which doesn't work. What _does_ match is `$dir/*\r$`. So CRLF line endings in GitHub essentially breaks subtrees when you use a squash merge strategy by default
